mlflow/evaluation.py

```diff
@@ -0,0 +1,224 @@
+from typing import Dict, Union
+import entrypoints
+import warnings
+import mlflow
+from mlflow.exceptions import MlflowException
+
+
+class EvaluationMetrics(dict):
+    pass
+
+
+class EvaluationArtifact:
+
+    @property
+    def content(self):
+        """
+        The content of the artifact (representation varies)
+        """
+        raise NotImplementedError()
+
+    @property
+    def location(self) -> str:
+        """
+        The location of the artifact
+        """
+        raise NotImplementedError()
+
+
+class EvaluationResult:
+
+    @classmethod
+    def load(cls, path):
+        """Load the evaluation results from the specified local filesystem path"""
+        raise NotImplementedError()
+
+    def save(self, path):
+        """Write the evaluation results to the specified local filesystem path"""
+        # We will likely avoid serializing artifacts themselves, just locations.
+        # Deserialization will resolve locations to artifact contents.
+        raise NotImplementedError()
+
+    @property
+    def metrics(self) -> EvaluationMetrics:
+        """
+        A dictionary mapping scalar metric names to scalar metric values
+        """
+        raise NotImplementedError()
+
+    @property
+    def artifacts(self) -> Dict[str, EvaluationArtifact]:
+        """
+        A dictionary mapping standardized artifact names (e.g. "roc_data") to
+        artifact content and location information
+        """
+        raise NotImplementedError()
+
+
+class EvaluationDataset:
+    """
+    Represents an input dataset for model evaluation. This is intended for
+    use with the `mlflow.evaluate()`API.
+    """
+
+    def __init__(self, data, labels=None, name=None):
+        """
+        :param data: One of the following:
+         - A numpy array or list of evaluation features, excluding labels.
+         - A Pandas DataFrame, or the path to a serialized DataFrame,
```

Comment:
Question:
For Pandas DataFrame case, does it mean we regards all columns as feature columns except the "labels" column ? @dbczumar

Reply:
A lot of DS teams put identification tags and additional information in the "raw input" that goes into an ML pipeline, filtering out the X (effective vector for Pandas ML use cases) columns before submitting to a processing pipeline and retain those identifying columns on batch inference for identification purposes (i.e. a member_id, a sku, a product_id, etc..) Most of these are either primary or surrogate keys and would be problematic to include in any evaluation of the feature set.

Reply:
I think we should leave it up to the user to filter out features that are unnecessary for model evaluation. @WeichenXu123 let's regard all columns except the `labels` column as features and document this so that users know they need to filter out everything that isn't a feature or a label.